### ISSUE TITLE
export setupPgClientTransaction()

### ISSUE THREAD

**Issue by wenzowski** [postgraphql/postgraphql#608]
Sometimes (eg. in tests) it's helpful to be able to use our own `pgClient` instance and transaction context instead of the one generated by `pgPool.connect()` in `withPostGraphQLContext()`. Can we export [`setupPgClientTransaction`](https://github.com/postgraphql/postgraphql/blob/master/src/postgraphql/withPostGraphQLContext.ts#L96) please?

**Comment by benjie** [comment]
You could just pass a custom pgPool into `withPostGraphQLContext`

```js
const myClient = await pool.connect();
augment(myClient);
await withPostGraphQLContect({
  pgPool: {
    connect: () => myClient,
  },
  ...
}, context => ...);
```

Note that a standard pg Client and a client from a pool are NOT identical - they handle errors differently for a start, which means that using a regular client (as opposed to a pgPool client) within async/await can result in promises never resolving when an error occurs. You can work around this, but it's a chore.

If this isn't sufficient, please expand upon what you need with a code example 👍

**Comment by wenzowski** [comment]
Gotcha, that's really important to know.

In the test example, I'm beginning a pg transaction, migrating up using an external library, running graphql tests within this transaction context, then rolling back the transaction. This is one of the strengths of postgres: parallel transactions allow a pristine database environment for each test by provisioning one connection per test.

Since the current connection already has an active transaction, we want to skip `await pgClient.query('begin')` and run `setupPgClientTransaction()` within the existing transaction context.

Here's a rough extraction of what I'm doing. In this example I've stubbed `withPostGraphQLContext()` to support a connection as the first arg:
```javascript
test('something', async () => {
  const rollback = new Error('rollback');
  try {
    knex.transaction(async (trx) => {
      await trx.migrate.latest();
      const pgClient = await trx.client.acquireConnection();
      const pgPublicSchema = 'app_schema';
      const pgOptions = {
        exportGqlSchemaPath: resolve(__dirname, './schema.graphql'),
        jwtSecret: config.get('secret'),
        jwtAudiences: 'config.get('audiences')',
        jwtPgTypeIdentifier: 'app_schema.jwt_token',
        pgDefaultRole: 'app_schema_anonymous',
      };
      const graphqlSchema = await createPostGraphQLSchema(
        pgClient, pgPublicSchema, pgOptions
      );
      const query = `
        {
          allThings {
            nodes {
              id
            }
          }
        }
      `;
      const result = await withPostGraphQLContext({pgClient, ...pgOptions},
        async (context) => {
          return await graphql(
            graphqlSchema,
            query,
            null,
            {...context},
            null,
            'allThings'
          );
        }
      )
      expect(result).toMatchSnapshot();

      throw rollback
    })
  } catch(error) {
    if (error === rollback) return;
    console.error(error && error.stack);
    throw error;
  } finally {
    knex.client.destroy();
  };
});
```

**Comment by wenzowski** [comment]
> using a regular client (as opposed to a pgPool client) within async/await can result in promises never resolving when an error occurs

Hmm, I think I've experienced this behaviour, external to postgraphql. Perhaps I should rewrite my migrations system to support [`pg-pool`](https://github.com/brianc/node-pg-pool) in place of [`generic-pool`](https://github.com/coopernurse/node-pool)...

**Comment by benjie** [comment]
Yeah; I've hit it a number of times over the last couple years. I now only use pgPool when using async/await, even if I only need one client.

I was wanting to tweak this myself actually; I want to be able to replace the `begin` with `begin isolation level serializable` and the `commit` with a `rollback` for tests also. I'd be open to a `beginStatement` and `commitStatement` option that can be passed (they could accept multiple statements too of course; so the `commitStatement` could be `commit; reset all;` for example).
